aws-datastore: Initial cloud sync failed.

I have set up full stack project for my Android app, generated model files, integrated in the app, and connected to cloud. I have enabled Cognito user pool for authorization, AppSync and S3 for storage. When running the app I am getting following exception:

```console
2020-06-17 16:57:56.404 12695-12989/   W/amplify:aws-datastore: Failed to sync down cloud state for model type = data
    AmplifyException {message=Failure performing sync query to AppSync: [GraphQLResponse.Error{message='Validation error of type FieldUndefined: Field 'syncdatas' in type 'Query' is undefined @ 'syncdatas'', locations='[GraphQLLocation{line='2', column='3'}]', path='null', extensions='null'}], cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$0$AppSyncClient(AppSyncClient.java:3)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$DAlM2A4r8A_52MV-BaSjc1SYWao.accept(Unknown Source:10)
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onResponse(AppSyncGraphQLOperation.java:16)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:504)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
2020-06-17 16:57:56.418 12695-12989/   W/amplify:aws-datastore: Initial cloud sync failed.
    AmplifyException {message=Failure performing sync query to AppSync: [GraphQLResponse.Error{message='Validation error of type FieldUndefined: Field 'syncdatas' in type 'Query' is undefined @ 'syncdatas'', locations='[GraphQLLocation{line='2', column='3'}]', path='null', extensions='null'}], cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$0$AppSyncClient(AppSyncClient.java:3)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$DAlM2A4r8A_52MV-BaSjc1SYWao.accept(Unknown Source:10)
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onResponse(AppSyncGraphQLOperation.java:16)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:504)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
```

I created a new project and did all configurations for amplify. I could successfully sign up a new user using my app and I could see new user created in my AWS Cognito user pool . Added new item to appsync DynamoDB with Amplify.DataStore.save call and S3 bucket using Amplify.Storage.uploadFile function call.Both data is saved only locally not syncing to AWS cloud.So I deleted the newly added data from app. and again logged into app.Then I get following exceptions again.


2020-06-25 16:56:30.540 6215-6365/com.prodigy.kospet W/amplify:aws-datastore: Failed to sync down cloud state for model type = Kospetdata
    AmplifyException {message=Failure performing sync query to AppSync: [GraphQLResponse.Error{message='Validation error of type FieldUndefined: Field 'syncKospetdatas' in type 'Query' is undefined @ 'syncKospetdatas'', locations='[GraphQLLocation{line='2', column='3'}]', path='null', extensions='null'}], cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$0$AppSyncClient(AppSyncClient.java:3)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$DAlM2A4r8A_52MV-BaSjc1SYWao.accept(Unknown Source:10)
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onResponse(AppSyncGraphQLOperation.java:16)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:504)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
2020-06-25 16:56:30.541 6215-6365/com.prodigy.kospet W/amplify:aws-datastore: Initial cloud sync failed.
    AmplifyException {message=Failure performing sync query to AppSync: [GraphQLResponse.Error{message='Validation error of type FieldUndefined: Field 'syncKospetdatas' in type 'Query' is undefined @ 'syncKospetdatas'', locations='[GraphQLLocation{line='2', column='3'}]', path='null', extensions='null'}], cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$0$AppSyncClient(AppSyncClient.java:3)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$DAlM2A4r8A_52MV-BaSjc1SYWao.accept(Unknown Source:10)
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onResponse(AppSyncGraphQLOperation.java:16)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:504)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)

2020-06-25 16:56:31.178 6215-6334/com.prodigy.kospet W/amplify:aws-datastore: Failed to publish a local change = PendingMutation{mutatedItem=Kospetdata {id=ac1c0a70-89a5-4bdf-abd9-8d8382a32621, title=kospetaudio, description=kospetaudio, audiopath=https://kospet511a0d8b7dee438a8bf514658de1f899162435-amplify.s3.amazonaws.com/public/kospetaudio, sensitivity=NORMAL, timer=fifteen}, classOfMutatedItem=class com.amplifyframework.datastore.generated.model.Kospetdata, mutationType=DELETE, mutationId=5c72bb5c-b724-11ea-bc89-9d5b66290272, predicate=null}
    AmplifyException {message=Wanted 1 metadata for item with id = ac1c0a70-89a5-4bdf-abd9-8d8382a32621, but had 0., cause=null, recoverySuggestion=This is likely a bug. please report to AWS.}
        at com.amplifyframework.datastore.syncengine.VersionRepository.extractVersion(VersionRepository.java:25)
        at com.amplifyframework.datastore.syncengine.VersionRepository.lambda$null$0$VersionRepository(VersionRepository.java:1)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$VersionRepository$R5kiVnQCzDy1jTk4Odil1m_dDtM.accept(Unknown Source:8)
        at com.amplifyframework.datastore.storage.sqlite.SQLiteStorageAdapter.lambda$query$4$SQLiteStorageAdapter(SQLiteStorageAdapter.java:42)
        at com.amplifyframework.datastore.storage.sqlite.-$$Lambda$SQLiteStorageAdapter$rGner20-W7t2tJu4yCrrZwlquWE.run(Unknown Source:10)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
2020-06-25 16:56:31.179 6215-6350/com.prodigy.kospet W/amplify:aws-datastore: Error ended observation of mutation outbox: 
    java.lang.RuntimeException: AmplifyException {message=Wanted 1 metadata for item with id = ac1c0a70-89a5-4bdf-abd9-8d8382a32621, but had 0., cause=null, recoverySuggestion=This is likely a bug. please report to AWS.}
        at io.reactivex.internal.util.ExceptionHelper.wrapOrThrow(ExceptionHelper.java:46)
        at io.reactivex.internal.observers.BlockingMultiObserver.blockingAwait(BlockingMultiObserver.java:185)
        at io.reactivex.Completable.blockingAwait(Completable.java:1256)
        at com.amplifyframework.datastore.syncengine.MutationProcessor.lambda$startDrainingMutationOutbox$1$MutationProcessor(MutationProcessor.java:6)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$MutationProcessor$iu5hAaGgWV8-jjRjPW1wTtFQOwM.apply(Unknown Source:4)
        at io.reactivex.internal.operators.observable.ObservableFlatMapCompletableCompletable$FlatMapCompletableMainObserver.onNext(ObservableFlatMapCompletableCompletable.java:97)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:201)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:255)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: AmplifyException {message=Wanted 1 metadata for item with id = ac1c0a70-89a5-4bdf-abd9-8d8382a32621, but had 0., cause=null, recoverySuggestion=This is likely a bug. please report to AWS.}
        at com.amplifyframework.datastore.syncengine.VersionRepository.extractVersion(VersionRepository.java:25)
        at com.amplifyframework.datastore.syncengine.VersionRepository.lambda$null$0$VersionRepository(VersionRepository.java:1)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$VersionRepository$R5kiVnQCzDy1jTk4Odil1m_dDtM.accept(Unknown Source:8)
        at com.amplifyframework.datastore.storage.sqlite.SQLiteStorageAdapter.lambda$query$4$SQLiteStorageAdapter(SQLiteStorageAdapter.java:42)
        at com.amplifyframework.datastore.storage.sqlite.-$$Lambda$SQLiteStorageAdapter$rGner20-W7t2tJu4yCrrZwlquWE.run(Unknown Source:10)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 

I did amplify delete and created a new project. Configured amplify. My android app could successfully create a new user in AWS cognito user pool and I added data which successfully uploaded to AWS appsync Dynamo DB and image got stored in AWS S3 bucket.Then I deleted data through my android app ( which called  Amplify.DataStore.delete ).Added new item through app.But deleted items are still there in Dynamo DB, new item didn't get added to AWS cloud.Also following exceptions happens which says "recoverySuggestion=This is likely a bug. please report to AWS"

 amplify:aws-datastore: Failed to sync down cloud state for model type = Kospetdata
    AmplifyException {message=Failure performing sync query to AppSync: [GraphQLResponse.Error{message='Validation error of type FieldUndefined: Field 'syncKospetdatas' in type 'Query' is undefined @ 'syncKospetdatas'', locations='[GraphQLLocation{line='2', column='3'}]', path='null', extensions='null'}], cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$0$AppSyncClient(AppSyncClient.java:3)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$DAlM2A4r8A_52MV-BaSjc1SYWao.accept(Unknown Source:10)
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onResponse(AppSyncGraphQLOperation.java:16)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:504)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
2020-06-25 20:49:14.018 16748-16949/com.prodigy.kospet W/amplify:aws-datastore: Initial cloud sync failed.
    AmplifyException {message=Failure performing sync query to AppSync: [GraphQLResponse.Error{message='Validation error of type FieldUndefined: Field 'syncKospetdatas' in type 'Query' is undefined @ 'syncKospetdatas'', locations='[GraphQLLocation{line='2', column='3'}]', path='null', extensions='null'}], cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$0$AppSyncClient(AppSyncClient.java:3)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$DAlM2A4r8A_52MV-BaSjc1SYWao.accept(Unknown Source:10)
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onResponse(AppSyncGraphQLOperation.java:16)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:504)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
2020-06-25 20:49:14.019 16748-16899/com.prodigy.kospet I/amplify:aws-datastore: Starting processing subscription data buffer.
2020-06-25 20:49:14.020 16748-16931/com.prodigy.kospet I/amplify:aws-datastore: Started processing the mutation outbox. Pending mutations will be published to the cloud.
2020-06-25 20:49:14.024 16748-16899/com.prodigy.kospet W/amplify:aws-datastore: Failed to publish a local change = PendingMutation{mutatedItem=Kospetdata {id=ac1c0a70-89a5-4bdf-abd9-8d8382a32621, title=kospetaudio, description=kospetaudio, audiopath=https://kospet511a0d8b7dee438a8bf514658de1f899162435-amplify.s3.amazonaws.com/public/kospetaudio, sensitivity=NORMAL, timer=fifteen}, classOfMutatedItem=class com.amplifyframework.datastore.generated.model.Kospetdata, mutationType=DELETE, mutationId=5c72bb5c-b724-11ea-bc89-9d5b66290272, predicate=null}
    AmplifyException {message=Wanted 1 metadata for item with id = ac1c0a70-89a5-4bdf-abd9-8d8382a32621, but had 0., cause=null, recoverySuggestion=This is likely a bug. please report to AWS.}
        at com.amplifyframework.datastore.syncengine.VersionRepository.extractVersion(VersionRepository.java:25)
        at com.amplifyframework.datastore.syncengine.VersionRepository.lambda$null$0$VersionRepository(VersionRepository.java:1)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$VersionRepository$R5kiVnQCzDy1jTk4Odil1m_dDtM.accept(Unknown Source:8)
        at com.amplifyframework.datastore.storage.sqlite.SQLiteStorageAdapter.lambda$query$4$SQLiteStorageAdapter(SQLiteStorageAdapter.java:42)
        at com.amplifyframework.datastore.storage.sqlite.-$$Lambda$SQLiteStorageAdapter$rGner20-W7t2tJu4yCrrZwlquWE.run(Unknown Source:10)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
2020-06-25 20:49:14.026 16748-16931/com.prodigy.kospet W/amplify:aws-datastore: Error ended observation of mutation outbox: 
    java.lang.RuntimeException: AmplifyException {message=Wanted 1 metadata for item with id = ac1c0a70-89a5-4bdf-abd9-8d8382a32621, but had 0., cause=null, recoverySuggestion=This is likely a bug. please report to AWS.}
        at io.reactivex.internal.util.ExceptionHelper.wrapOrThrow(ExceptionHelper.java:46)
        at io.reactivex.internal.observers.BlockingMultiObserver.blockingAwait(BlockingMultiObserver.java:185)
        at io.reactivex.Completable.blockingAwait(Completable.java:1256)
        at com.amplifyframework.datastore.syncengine.MutationProcessor.lambda$startDrainingMutationOutbox$1$MutationProcessor(MutationProcessor.java:6)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$MutationProcessor$iu5hAaGgWV8-jjRjPW1wTtFQOwM.apply(Unknown Source:4)
        at io.reactivex.internal.operators.observable.ObservableFlatMapCompletableCompletable$FlatMapCompletableMainObserver.onNext(ObservableFlatMapCompletableCompletable.java:97)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:201)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:255)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: AmplifyException {message=Wanted 1 metadata for item with id = ac1c0a70-89a5-4bdf-abd9-8d8382a32621, but had 0., cause=null, recoverySuggestion=This is likely a bug. please report to AWS.}
        at com.amplifyframework.datastore.syncengine.VersionRepository.extractVersion(VersionRepository.java:25)
        at com.amplifyframework.datastore.syncengine.VersionRepository.lambda$null$0$VersionRepository(VersionRepository.java:1)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$VersionRepository$R5kiVnQCzDy1jTk4Odil1m_dDtM.accept(Unknown Source:8)
        at com.amplifyframework.datastore.storage.sqlite.SQLiteStorageAdapter.lambda$query$4$SQLiteStorageAdapter(SQLiteStorageAdapter.java:42)
        at com.amplifyframework.datastore.storage.sqlite.-$$Lambda$SQLiteStorageAdapter$rGner20-W7t2tJu4yCrrZwlquWE.run(Unknown Source:10)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 

--- a/aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java
+++ b/aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java
@@ -91,15 +91,22 @@ final class AppSyncRequestFactory {
             throws DataStoreException {
 
         final StringBuilder doc = new StringBuilder();
-        final String capitalizedModelName = Casing.capitalizeFirst(modelClass.getSimpleName());
+
+        ModelSchema schema;
+        try {
+            schema = ModelSchema.fromModelClass(modelClass);
+        } catch (AmplifyException amplifyException) {
+            throw new DataStoreException("Failed to get fields for model.",
+                    amplifyException, "Validate your model file.");
+        }
+        final String capitalizedPluralName = Casing.capitalizeFirst(schema.getPluralName());
 
         int indent = 0;
         // Outer container, e.g. query SyncBlogPost {
-        doc.append("query Sync").append(capitalizedModelName).append("s {\n");
+        doc.append("query Sync").append(capitalizedPluralName).append(" {\n");
 
         // Inner directive, e.g. syncBlogPosts(
-        doc.append(padBy(++indent))
-            .append("sync").append(capitalizedModelName).append("s");
+        doc.append(padBy(++indent)).append("sync").append(capitalizedPluralName);
 
         // Optional param for inner directive, i.e. (lastSync: 123123)
         // (lastSync:11123123, nextToken: "asdfasdfaS")

